move away from the `FromPyObject` blanket for `PyClass + Clone`

Icxolu opened the pull request:
This proposes a first step in removing the `FromPyObject` blanket implementation for cloneable pyclasses.

This adds a marker trait `ExtractPyClassWithClone` which is automatically implemented by the `#[pyclass]` macro for every pyclass and modifies the blanket to require the marker. I think this should be functionally equivalent. It also adds a new `skip_from_py_object` option to opt out of emitting the marker trait and thus enable a custom `FromPyObject` implementation.

The following steps could then be:
- add an option to explicitly opt-in (this can then just emit the `FromPyObject` impl directly)
- emit a warning for pyclasses that do not declare out-out nor opt in 
- switch the default from opt-out to opt-in
- remove the blanket and the marker and deprecate the out-out option


Ref #5419

davidhewitt commented:
This approach looks good to me, thanks. Documentation with the rest of the `#[pyclass]` options needs updating, and somewhere in the guide probably too, maybe in the "Conversion Traits" section?

Icxolu commented:
Thanks for the feedback! I've added the newsfragment, extended the pyclass option table and added a paragraph in the guide. Should we also mention this in the migration guide already, are do we wait until we emit a warning?

davidhewitt commented on guide/pyclass-parameters.md:
```diff
@@ -21,6 +21,7 @@
 | `rename_all = "renaming_rule"` | Applies renaming rules to every getters and setters of a struct, or every variants of an enum. Possible values are: "camelCase", "kebab-case", "lowercase", "PascalCase", "SCREAMING-KEBAB-CASE", "SCREAMING_SNAKE_CASE", "snake_case", "UPPERCASE". |
 | `sequence` |  Inform PyO3 that this class is a [`Sequence`][params-sequence], and so leave its C-API mapping length slot empty. |
 | `set_all` | Generates setters for all fields of the pyclass. |
+| `skip_from_py_object` | Prevents this PyClass from participarting in the `FromPyObject: PyClass + Clone` blanket implementation. This allows a custom `FromPyObject` impl, even if `self` is `Clone`. |
```
```suggestion
| `skip_from_py_object` | Prevents this PyClass from participating in the `FromPyObject: PyClass + Clone` blanket implementation. This allows a custom `FromPyObject` impl, even if `self` is `Clone`. |
```

davidhewitt reviewed (approved):
Perfect, thanks!

I think no need for this to go in the migration guide until we begin emitting the warning. Users don't need to do anything with this to upgrade to 0.27.
